### ISSUE TITLE
smvJoinByKey "Semi" Join not working

### ISSUE THREAD

**Issue by XiaosenChen** [TresAmigosSD/SMV#603]
In Python Interface, this is the error I got:
```
/home/cchen/SMV/src/main/python/smv/helpers.pyc in method(self, *args)
    173 def _getUnboundMethod(helperCls, methodName):
    174     def method(self, *args):
--> 175         return getattr(helperCls(self), methodName)(*args)
    176     return method
    177 

/home/cchen/SMV/src/main/python/smv/helpers.pyc in smvJoinByKey(self, other, keys, joinType)
    216     # workaround
    217     def smvJoinByKey(self, other, keys, joinType):
--> 218         jdf = self._jPythonHelper.smvJoinByKey(self._jdf, other._jdf, _to_seq(keys), joinType)
    219         return DataFrame(jdf, self._sql_ctx)
    220 

/usr/local/spark/current/python/lib/py4j-0.8.2.1-src.zip/py4j/java_gateway.py in __call__(self, *args)
    536         answer = self.gateway_client.send_command(command)
    537         return_value = get_return_value(answer, self.gateway_client,
--> 538                 self.target_id, self.name)
    539 
    540         for temp_arg in temp_args:

/usr/local/spark/current/python/pyspark/sql/utils.py in deco(*a, **kw)
     40                 raise AnalysisException(s.split(': ', 1)[1])
     41             if s.startswith('java.lang.IllegalArgumentException: '):
---> 42                 raise IllegalArgumentException(s.split(': ', 1)[1])
     43             raise
     44     return deco

IllegalArgumentException: Unsupported join type 'Semi'. Supported join types include: 'inner', 'outer', 'full', 'fullouter', 'leftouter', 'left', 'rightouter', 'right', 'leftsemi'.
```

**Comment by ninjapapa** [comment]
According to Spark's doc:
http://spark.apache.org/docs/2.1.0/api/python/pyspark.sql.html?highlight=join#pyspark.sql.DataFrame.join

"semi" is not a supported join type. There is a "leftsemi" instead. 

Since our smvJoinByKey eventually call Spark's join, the supported join types are limited to whatever Spark supported.

**Comment by XiaosenChen** [comment]
I am using smvJoinbyKey with "Semi" here, it seems it gets converted to Spark leftsemi here:
```
  object SmvJoinType {
    val Inner = "inner"
    val Outer = "outer"
    val LeftOuter = "leftouter"
    val RightOuter = "rightouter"
    val Semi = "leftsemi"
  }
```
But I got this error trace.

**Comment by ninjapapa** [comment]
There are no `SmvJoinType` equivalent on python side, so the `JoinType` parameter of the Python `smvJoinByKey` is a pure string.

**Comment by AliTajeldin** [comment]
Perhaps we should create the python consts equivalent to `SmvJoinType` to get an earlier error on typos (still runtime error but during the compile of the python code rather during run of spark).

**Comment by ninjapapa** [comment]
JoinType doesn't exist error is not too bad, since it happens when building the logical plan. It is run time, but no data processing yet. On the other hand, for new users, introducing the `SmvJoinType` is another thing they need to learn. May not worth it.

**Comment by XiaosenChen** [comment]
Based on previous error trace, I passed string "leftsemi" into smvJoinByKey, get this error, which I really don't understand. On the other hand, passing "leftsemi" into Spark's join works.
```
---------------------------------------------------------------------------
IllegalArgumentException                  Traceback (most recent call last)
<ipython-input-72-bb354658529c> in <module>()
----> 1 Dig.smvRenameField(("DIAG_CDE", "CONCEPT_CODE")).smvJoinByKey(DigVocab, ["CONCEPT_CODE"], "leftsemi").show()

/home/cchen/SMV/src/main/python/smv/helpers.pyc in method(self, *args)
    173 def _getUnboundMethod(helperCls, methodName):
    174     def method(self, *args):
--> 175         return getattr(helperCls(self), methodName)(*args)
    176     return method
    177 

/home/cchen/SMV/src/main/python/smv/helpers.pyc in smvJoinByKey(self, other, keys, joinType)
    216     # workaround
    217     def smvJoinByKey(self, other, keys, joinType):
--> 218         jdf = self._jPythonHelper.smvJoinByKey(self._jdf, other._jdf, _to_seq(keys), joinType)
    219         return DataFrame(jdf, self._sql_ctx)
    220 

/usr/local/spark/current/python/lib/py4j-0.8.2.1-src.zip/py4j/java_gateway.py in __call__(self, *args)
    536         answer = self.gateway_client.send_command(command)
    537         return_value = get_return_value(answer, self.gateway_client,
--> 538                 self.target_id, self.name)
    539 
    540         for temp_arg in temp_args:

/usr/local/spark/current/python/pyspark/sql/utils.py in deco(*a, **kw)
     40                 raise AnalysisException(s.split(': ', 1)[1])
     41             if s.startswith('java.lang.IllegalArgumentException: '):
---> 42                 raise IllegalArgumentException(s.split(': ', 1)[1])
     43             raise
     44     return deco

IllegalArgumentException: requirement failed
```

**Comment by XiaosenChen** [comment]
I think this may actually be a long surviving bug in smvJoinByKey, I can look into it.

**Comment by ninjapapa** [comment]
I think I figured out the problem.
Since `leftsemi` actually do a projection after the join operation (only return columns from the left table), the following step:
```scala
dfCoalescedKeys.smvSelectMinus(rightKeys(0), rightKeys.tail: _*)
```
caused the issue. 

Need to change it to
```scala
if (joinType == SmvJoinType.Semi) dfCoalescedKeys
else dfCoalescedKeys.smvSelectMinus(rightKeys(0), rightKeys.tail: _*)
```

**Comment by XiaosenChen** [comment]
@ninjapapa you stole my only ticket...

**Comment by laneb** [comment]
@AliTajeldin Also we wouldn't be guaranteed an error during Python compile. It how users import it. E.g.
```
from smv.SmvJoinType import Semi 
```
would cause an `ImportError` at run time, but 
```
from smv import SmvJoinType
class X(SmvModule):
...
  def run(self, i):
    return i[Y].smvJoin(i[Z], "mykey", SmvJoinType.Semi)
```
would cause an `AttributeError` at runtime.

**Comment by ninjapapa** [comment]
@XiaosenChen ticket is still yours :-)

**Comment by AliTajeldin** [comment]
@laneb correct.  But it would be a run time error at the time we are creating the graph not when we execute it as part of the spark action.  Closer to the root cause and should be easier to detect the root cause instead of the stacktrace shown in the original message.

**Comment by XiaosenChen** [comment]
fixed 6de2a89742bb3df066c10685da9aee8f5c1d084d
